New metrics version with aggregation of InfluxDB metrics data

sixtedemaupeou opened the pull request:
Fix https://github.com/etalab/data.gouv.fr/issues/915

- Daily aggregation of InfluxDB data to count the number of resource, dataset, organisation and reuse views.
- Deletion of unaggregated metrics data in InfluxDB.
- Aggregation of daily metrics into all-time metrics and storage of these in udata MongoDB.

Foreseeable issues are:
- Vector not detecting new log files?
- udata-metrics job aggregate-metrics-last-day running before Vector has completed.
- Missing community resources and other models that were counted by udata-piwik.
- Entering metrics data in Mongo may take too much time (~1h at the moment).

sixtedemaupeou committed:
WIP update to build metrics from Vector instead of udata-piwik (Matomo)

sixtedemaupeou committed:
replace influxdb package with influxdb-client which supports Flux and InfluxDB 2.0

sixtedemaupeou committed:
turn aggregation into a job and document in readme

sixtedemaupeou committed:
fix dependency issue and update Influx aggregation query

sixtedemaupeou committed:
fix aggregation query and increase timeout

sixtedemaupeou committed:
fix logs deletion query

sixtedemaupeou committed:
setup aggregation of Influx views per model and write to Mongo

sixtedemaupeou committed:
uncomment query code and replace query date

sixtedemaupeou committed:
minor comment change

sixtedemaupeou committed:
rename yield in query

maudetes commented on requirements/install.pip:
```diff
@@ -1,2 +1,2 @@
-udata>=2.0.0
-influxdb==5.2.3
+udata==4.0.3.dev0
+influxdb-client==1.26.0
```
We could use `1.32.0` already?

sixtedemaupeou replied:
I wanted to but this would require updating udata as well as there is a conflict with urllib version in that case

maudetes commented on requirements/install.pip:
```diff
@@ -1,2 +1,2 @@
-udata>=2.0.0
-influxdb==5.2.3
+udata==4.0.3.dev0
```
We shouldn't ping a dev version
```suggestion
udata>=4.0.0
```

maudetes commented on udata_metrics/tasks.py:
```diff
@@ -1,29 +1,20 @@
     }
-    client.insert(body)
+    for _, measurement in measurements_dict.items():
+        client = metrics_client_factory()
```
Could we create client only once?

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
+        self.client = InfluxDBClient(**dsn, timeout=600_000)
+
+    def aggregate_metrics(self, measurement: str):
+        """measurement one of reuse.reuse_id, resource.resource_id,
```
```suggestion
        """measurement is one of reuse.reuse_id, resource.resource_id,
```

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
+        yesterday = yesterday_dt.strftime("%Y-%m-%d")
+        page_type = f"{measurement}".split(".")[0]
+        log.info(f"Running metrics aggregation for {page_type}")
+        id_key = f"{measurement}".split(".")[1]
```
```suggestion
        page_type, id_key = measurement.split(".")
        log.info(f"Running metrics aggregation for {page_type}")
```

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
+        page_type = f"{measurement}".split(".")[0]
+        log.info(f"Running metrics aggregation for {page_type}")
+        id_key = f"{measurement}".split(".")[1]
+        # TODO: Replace with right timezone for logs in the query
```
Don't forget TODO

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
+        # TODO: Replace with right timezone for logs in the query
+        agg_query = f"""
+                import "date"
+                from(bucket: "vector-bucket")
```
I guess we should use METRICS_VECTOR_BUCKET?

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
+        result = self.client.query_api().query(retrieve_metrics_query)
+        for table in result:
+            for record in table:
+                # Currently we ignore resource hits as they are different from resource downloads.
```
A resource hit is a call to the API `api/1/datasets/<dat_id>/resources/<res_id>` or `api/2/datasets/resources/<res_id>` and not a download right?

sixtedemaupeou replied:
Yes, except I did not know we could have something like `api/2/datasets/resources/<res_id>`. Is it correct? Here are the regexes we are using in [vector-test](https://github.com/sixtedemaupeou/vector-test)

resource_id_match: `r'^(https://www.data.gouv.fr)?/(?P<language_or_api>\w{2}|api/[1-2])/datasets/r/(?P<resource_id>[a-zA-Z0-9-]+)'`
resource_hit_match:` r'^(https://www.data.gouv.fr)?/(?P<api_version>api/[1-2])/datasets/(?P<dataset_id_or_slug>[a-zA-Z0-9-]+)/resources/(?P<resource_id>[a-zA-Z0-9-]+)/'`

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
-        result = self.client.query(query)
-        return result
+                    model_id = record[id_key]
+                    model_result = model.objects.filter(id=model_id).first()
```
```suggestion
                    model_result = model.get(model_id)
```

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
-        self.client = InfluxDBClient(**dsn)
+        self.client = InfluxDBClient(**dsn, timeout=600_000)
+
+    def aggregate_metrics(self, measurement: str):
```
Could we maybe split this function in multiple functions, since there are 3 different main parts?

maudetes reviewed (commented):
Nice! :clap:

I've made some comments but most are marginal.
I could not test its integration with the vector/influx pipeline, thus I haven't tested the code yet.
I would also like to explore the queries more when I have some test data :)

I wonder if we could add some test on this new aggregation, but not sure how since most of the actual logic is in queries sent to influx. We could probably test the metrics update in udata models with a mock.

maudetes commented on udata_metrics/client.py:
```diff
@@ -1,26 +1,104 @@
+                    # treat it separately.
+                    try:
+                        model_id = uuid.UUID(record[id_key])
+                        model_result = get_resource(model_id)
```
Sadly metric aggregation takes too much time, probably due to `get_resource` taking a significant time * the large number of resources.

maudetes replied:
After a bit of inquiry, get_resource can take a lot of time if the targeted resource is in a dataset with numerous resources.
We've tried optimizing the get_resource without success. For info:
- having the dataset id, getting the dataset first before getting the precise resource, but it was not efficient enough either
- querying only relevant resource from Mongo `Dataset.objects.filter(resources__id=id).fields(resources={'$elemMatch': {'_id': str(id)}}).first()`, it would not work at validation time, since the entire dataset hasn't been returned.

However, it appeared that the `.save()` command would also take a big amount of time (more than fetching) for those huge datasets.

The other solution is to iterate on dataset x resources objects directly, updating the metric only if found in the table, and saving the dataset once only. An example solution has been pushed here: https://github.com/opendatateam/udata-metrics/commit/d4f644c7d1a4dd84de610be2aefed9d5e72f540e.
The code is ugly, but shows a proposal with an update of resources metrics running under 5mins locally on my computer.

sixtedemaupeou committed:
Update udata_metrics/client.py mongo query

Co-authored-by: maudetes <maudet.estelle@gmail.com>

sixtedemaupeou committed:
minor improvements

maudetes committed:
Add minimal .gitignore

maudetes committed:
Remove circle ci filter on branch

maudetes committed:
Add tests on metrics updates in udata models

maudetes committed:
Add pytest-mock deps

maudetes committed:
Split aggregation logic in functions

maudetes committed:
Refactor resources model update

maudetes committed:
Add max-line-length in flake

maudetes committed:
Add edgy cases in client test

- Metrics for object that don't exist in the db
- some DB objects that don't have any metrics computed

maudetes committed:
Fix typo

maudetes committed:
Update Readme

maudetes commented:
I'm converting this to a draft, so that we don't merge it by mistake before we're ready to.

Some comments for when we're back on this topic.

> Foreseeable issues are:
> Vector not detecting new log files?

We've tested but it does detect new log files. The file detection is also configurable.

> udata-metrics job aggregate-metrics-last-day running before Vector has completed.

Question still open : we can just make sure to schedule after a large enough time?

> Missing community resources and other models that were counted by udata-piwik.

Not blocking for now and can probably be updated with the same logic if needed.

> Entering metrics data in Mongo may take too much time (~1h at the moment).

It actually took way longer with the prod database. It has been improved in the latest commits, leading to 10 min~ on some hours of logs. We should benchmark it on an entire day log.

Other things to consider:
We could remove the need to download logs from s3 if logs were already pushed at the right place from cron.
We should make sure that udata-metrics does not fail silently
We should check the behaviour if a log file is feeded twice to influx. Would the aggregation step double the values?

maudetes commented:
Replaced by https://github.com/opendatateam/udata-metrics/pull/8
